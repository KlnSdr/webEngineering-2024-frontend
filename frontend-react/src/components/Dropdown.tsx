import React, { useState } from "react";

interface DropdownProps {
  options: string[];
  initialValue: string;
  onChange: (value: string) => void;
}

const Dropdown: React.FC<DropdownProps> = ({
  options,
  initialValue,
  onChange,
}) => {
  const [value, setValue] = useState(initialValue);
  return (
    <select
      onChange={(e) => {
        setValue(e.target.value);
        onChange(e.target.value);
      }}
    >
      {options.map((option: string) => {
        if (option === value) {
          return (
            <option value={option} selected>
              {option}
            </option>
          );
        } else {
            return <option value={option}>{option}</option>;
        }
      })}
    </select>
  );
};

export default Dropdown;
