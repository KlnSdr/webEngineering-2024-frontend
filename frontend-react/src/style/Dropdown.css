.dropdownOptionsContainer {
    position: absolute;
    background-color: white;
    border: 1px solid black;
    width: 100%;
    max-height: 200px;
    overflow-y: scroll;
    max-width: 10vw;
    z-index: 9001;
    border-radius: 5px;
}

.dropdownOption:hover {
    background-color: #f1f1f1;
}
