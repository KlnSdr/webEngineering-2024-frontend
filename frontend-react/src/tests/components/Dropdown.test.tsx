import React, {act} from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import '@testing-library/jest-dom/extend-expect';
import Dropdown from "../../components/Dropdown";
import userEvent from "@testing-library/user-event"; // Adjust the import if the path is different

describe("Dropdown Component", () => {
    const options = ["Option 1", "Option 2", "Option 3"];

    test("renders with correct initial value", () => {
        const initialValue = "Option 2";
        const handleChange = jest.fn();

        const {asFragment} = render(<Dropdown options={options} initialValue={initialValue} onChange={handleChange} />);

        const selectElement = screen.getByRole("textbox");
        expect(selectElement).toHaveValue(initialValue);

        expect(asFragment()).toMatchSnapshot();
    });

    test("calls onChange callback when value changes", () => {
        const handleChange = jest.fn();

        render(<Dropdown options={options} initialValue={""} onChange={handleChange} />);

        const selectElement = screen.getByRole("textbox");
        userEvent.type(selectElement, "Option 3");
        screen.findByText("Option 3").then((option) => {
            console.log(option);
            fireEvent.click(option);
            expect(selectElement).toHaveValue("Option 3");
            expect(handleChange).toHaveBeenCalledTimes(1);
            expect(handleChange).toHaveBeenCalledWith("Option 3");
        });
    });
});
