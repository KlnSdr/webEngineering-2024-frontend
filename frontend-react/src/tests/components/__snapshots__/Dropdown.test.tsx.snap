// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown Component renders with correct initial value and options 1`] = `
<DocumentFragment>
  <select>
    <option
      value="Option 1"
    >
      Option 1
    </option>
    <option
      value="Option 2"
    >
      Option 2
    </option>
    <option
      value="Option 3"
    >
      Option 3
    </option>
  </select>
</DocumentFragment>
`;
